posts/2022-10-04-microprofile-serverless-ibm-code-engine.adoc

```diff
@@ -0,0 +1,412 @@
+
+Congratulations! You just called your application on IBM Cloud Code Engine.
+
+ If you open the IBM Cloud Code Engine UI to your application's **Overview** tab, you can see the active instances while you are making requests. You will use this page when you start making requests to observe IBM Cloud Code Engine creating the application instance and then scaling down to zero when it stops receiving work.
+
+image:/img/blog/ibm-code-engine-active-instances.png[IBM Cloud Code Engine Active Instances,width=70%,align="center"]
+
+A common use case for serverless applications is to process events coming from non-HTTP sources, such as Kafka topics or IBM Cloud Object Storage. Historically, to consume these events the application had to make use of the cloud provider's SDK, which locks the application into that provider. This is where CloudEvents comes in, as it provides a cross-provider standard around which data can be transmitted and received by an application, improving portability and reducing large dependencies.
+
+
+=== Updating the application to use CloudEvents
+IBM Cloud Code Engine connects IBM Cloud event providers such as https://cloud.ibm.com/docs/codeengine?topic=codeengine-working-kafkaevent-producer#setup-kafka-receiverapp[Kafka] or https://cloud.ibm.com/docs/codeengine?topic=codeengine-eventing-cosevent-producer#obstorage_ev_app[IBM Cloud Object Storage] and the application by using CloudEvents. These events are sent to a subscribed application as HTTP POST requests. Because the requests are in HTTP format, a RESTful application can receive these events without needing additional libraries and configuration.
+
+To include the CloudEvents `restful-ws` library within the application, Add the following https://openliberty.io/guides/maven-intro.html[Maven] dependency to your `pom.xml`:
+
+```
+<dependency>
+    <groupId>io.cloudevents</groupId>
+    <artifactId>cloudevents-http-restful-ws-jakarta</artifactId>
+    <version>2.5.0</version>
+</dependency>
+```
+
+Or if you use https://openliberty.io/guides/gradle-intro.html[Gradle]:
+```
+dependencies {
+	implementation group: 'io.cloudevents', name: 'cloudevents-http-restful-ws', version: '2.5.0'}
+
+```
+
+Two `restful-ws` packages are provided for CloudEvents:
+
+* `restful-ws` supports Jakarta EE 8 and the `javax.*` namespace
+* `restful-ws-jakarta` supports Jakarta EE 9+ and the `jakarta.*` namespace*
+
+The CloudEvent 2.5.0 release, the first release of the `restful-ws-jakarta` package, is at the time of this writing  not  yet available. If you want to use the new package before its release, you can build the package locally and store it in a suitable Maven repository for use with your application build. If you build from source,  change the version  to `2.5.0-SNAPSHOT`. You can find the source at https://github.com/cloudevents/sdk-java/tree/master/http/restful-ws-jakarta[HTTP Protocol Binding for Jakarta EE9+ - Jakarta RESTful Web Services].
```

Comment:
this part is confusing- I'm not sure Im reading it right but it seems like first we tell them to get version 2.5.0 as a dependency from Maven, but then its says that 2.5.0 isn't available yet and the need to get it from GH and upload to a Maven repo themselves? Is there another version they can use that is available? Or do they have to have to the jakarta package to complete the task?

Reply:
So we recently merged the code changes in to CloudEvents SDK main and we do not yet have a date for 2.5.0 release. the idea here was that when 2.5.0 is released, we would update the blog by simply removing this section. 

The idea is that they can still achieve the task with EE 9+ despite the package not yet being publically available, by building locally and having built locally what other changes they would need to make.

The section about the packages to use is to cover if they wanted to use EE 8 because that is what there App was using then the information they need is there, though would probably require a different Note to cover the change of version to 2.4.0

Reply:
The alternative package is applicable to MicroProfile 4.0. Microprofile 5+ requires the new package. To get that to work with 4.0 I think requires updating 5 files. So i will work on simplifying that part on just focussing on build locally, and can then update later when the package is formally released.